src/components/forms/pool_actions/MigrateForm/composables/useMigrateMath.ts

```diff
@@ -0,0 +1,124 @@
+
+import useNumbers from '@/composables/useNumbers';
+import useTokens from '@/composables/useTokens';
+import useUserSettings from '@/composables/useUserSettings';
+
+import { FullPool } from '@/services/balancer/subgraph/types';
+import PoolCalculator from '@/services/pool/calculator/calculator.sevice';
+
+import { bnSum, bnum } from '@/lib/utils';
+import useSlippage from '@/composables/useSlippage';
+
+export type MigrateMathResponse = ReturnType<typeof useMigrateMath>;
+
+export default function useMigrateMath(
+  fromPool: Ref<FullPool>,
+  toPool: Ref<FullPool>
+) {
+  /**
+   * COMPOSABLES
+   */
+  const { tokens, balances, balanceFor, getToken } = useTokens();
+  const { currency } = useUserSettings();
+  const { fNum, toFiat } = useNumbers();
+  const { minusSlippageScaled } = useSlippage();
+
+  /**
+   * SERVICES
+   */
+  const fromPoolCalculator = new PoolCalculator(
+    fromPool,
+    tokens,
+    balances,
+    'exit',
+    ref(false)
+  );
+
+  const toPoolCalculator = new PoolCalculator(
+    toPool,
+    tokens,
+    balances,
+    'join',
+    ref(false)
+  );
+
+  /**
+   * COMPUTED
+   */
+  const bptBalance = computed(() => balanceFor(fromPool.value.address));
+
+  const hasBpt = computed(() => bnum(bptBalance.value).gt(0));
+
+  const tokenCount = computed(() => fromPool.value.tokenAddresses.length);
+
+  const poolDecimals = computed(() => fromPool.value.onchain.decimals);
+
+  const bptBalanceScaled = computed(() =>
+    parseUnits(bptBalance.value, poolDecimals.value).toString()
+  );
+
+  const withdrawalTokens = computed(() =>
+    fromPool.value.tokenAddresses.map(address => getToken(address))
+  );
+
+  const fullAmounts = computed(() => {
+    const { receive } = fromPoolCalculator.propAmountsGiven(
+      bptBalance.value,
+      0,
+      'send'
+    );
+    return receive;
+  });
+
+  const fullAmountsScaled = computed((): string[] =>
+    fullAmounts.value.map((amount, i) =>
+      parseUnits(amount, withdrawalTokens.value[i].decimals).toString()
+    )
+  );
+
+  const swapAmountsOut = computed(() =>
+    fullAmountsScaled.value.map(minusSlippageScaled)
+  );
+
+  const fiatAmounts = computed((): string[] =>
+    fromPool.value.tokenAddresses.map((address, i) =>
+      toFiat(fullAmounts.value[i], address)
+    )
+  );
+
+  const fiatTotal = computed(() => bnSum(fiatAmounts.value).toString());
+
+  const fiatTotalLabel = computed(() => fNum(fiatTotal.value, currency.value));
+
+  const fullBPTOut = computed(() =>
+    fromPoolCalculator.exactTokensInForBPTOut(fullAmounts.value).toString()
+  );
+
+  const priceImpact = computed(() => {
+    // TODO: When from/to pool token count is different its not possible to calculate the price impact.
+    if (fromPool.value.tokensList.length !== toPool.value.tokensList.length) {
+      return 0;
+    }
+    return toPoolCalculator
+      .priceImpact(fullAmounts.value, {
+        queryBPT: fullBPTOut.value
+      })
+      .toNumber();
```

Comment:
@garethfuller does this calculation make sense? I'm using the `toPoolCalculator` here and grabbing the amounts from the `toPool`.

Reply:
The `fullBPTOut` passed in here is incorrect. This needs to be the bptOut from the bb-a-USD pool. At the moment you're using `exactTokensInForBPTOut` which is the bpt out if you were investing the stables back into staBal3.

I'm not sure if you have the function from the SDK here, but you need to use a query function which essentially simulates the whole batch relayer transaction and gives you the result. Please take to @johngrantuk about this if you're not sure where it comes from.